File: backend/src/main/java/com/esmo/empaas/services/BuildpackService.java
package com.esmo.empaas.services;

import java.util.List;
import java.util.Map;

import org.springframework.stereotype.Service;

import com.esmo.empaas.entities.BuildpackEntity;
import com.esmo.empaas.exceptions.ResourceAlreadyExistsException;
import com.esmo.empaas.exceptions.ResourceNotFoundException;
import com.esmo.empaas.repositories.BuildpackRepository;
import com.esmo.empaas.utils.DockerUtil;
import com.esmo.empaas.utils.FileUtil;

@Service
public class BuildpackService {

	private final FileUtil fileUtil;
	private final DockerUtil dockerUtil;
	private final BuildpackRepository buildpackRepository;

	public BuildpackService(FileUtil fileUtil, DockerUtil dockerUtil, BuildpackRepository buildpackRepository) {
		this.fileUtil = fileUtil;
		this.dockerUtil = dockerUtil;
		this.buildpackRepository = buildpackRepository;
	}

	public BuildpackEntity findById(String id) {
		return buildpackRepository.findById(id).orElseThrow(() -> new ResourceNotFoundException("Buildpack", "id", id));
	}

	public List<BuildpackEntity> findAll() {
		return buildpackRepository.findAll();
	}

	public BuildpackEntity select(String projectDir) {
		BuildpackEntity buildpack = null;

		if (fileUtil.searchFile(projectDir, "package.json")) {
			buildpack = buildpackRepository.findByName("nodejs")
					.orElseThrow(() -> new ResourceNotFoundException("Buildpack", "name", "nodejs"));
		}

		return buildpack;
	}

	public BuildpackEntity findByName(String name) {
		return buildpackRepository.findByName(name)
				.orElseThrow(() -> new ResourceNotFoundException("Buildpack", "name", name));
	}

	public BuildpackEntity save(BuildpackEntity data) {
		if (buildpackRepository.findByName(data.getName()).isPresent()) {
			throw new ResourceAlreadyExistsException("Buildpack", "name", data.getName());
		}

		return buildpackRepository.save(data);
	}

	public void build(String baseImage, String workdir, int port, String installCommand, String buildCommand,
			String startCommand, String baseDirectory, Map<String, String> envs) {
		switch (baseImage.split(":")[0]) {
		case "nodejs":
			dockerUtil.createNodeJsDockerFile(baseImage, workdir, port, installCommand, buildCommand, startCommand,
					baseDirectory, envs);

			break;
		default:
			break;
		}
	}
}

File: old/apps/backend/src/modules/buildpack/buildpack.service.ts
import { InjectRepository } from '@mikro-orm/nestjs';
import { EntityManager, EntityRepository } from '@mikro-orm/core';
import { Injectable, NotFoundException } from '@nestjs/common';

import { Buildpack } from './buildpack.entity';
import { searchFile } from 'src/shared/utils/file.util';
import createNodeDockerfile from 'src/shared/buildpacks/node';
import { BuildpackBuildRequestDto } from './buildpack.dto';
import { executeCommand } from 'src/shared/utils/exec.util';

@Injectable()
export class BuildpackService {
  constructor(
    @InjectRepository(Buildpack)
    private repository: EntityRepository<Buildpack>,
    private em: EntityManager,
  ) {}

  async findAll(): Promise<Buildpack[]> {
    return this.repository.findAll();
  }

  async findByName(name: string): Promise<Buildpack> {
    const buildpack = await this.repository.findOne({ name });
    if (!buildpack) {
      throw new NotFoundException('Buildpack not found');
    }

    return buildpack;
  }

  async selectBuildPack(projectDir: string): Promise<Buildpack> {
    let buildpack: Buildpack;

    if (searchFile(projectDir, 'package.json')) {
      buildpack = await this.findByName('nodejs');
    }

    return buildpack;
  }

  async build(data: BuildpackBuildRequestDto): Promise<void> {
    // TODO: select buildback based on language
    let buildpack: string;
    if (searchFile(data.projectDir, 'package.json')) {
      buildpack = 'nodejs';
      this.repository;
    }

    switch (buildpack) {
      case 'nodejs':
        createNodeDockerfile({
          baseDirectory: data.baseDirectory,
          baseImage: `${data.image}:${data.version}`,
          buildCommand: data.buildCommand,
          installCommand: data.installCommand,
          startCommand: data.startCommand,
          workdir: data.projectDir,
          port: data.port,
          envs: data.env,
        });

        executeCommand(`docker build -t :v1.0.0 .`, data.projectDir);

        break;
      default:
        break;
    }
  }
}
